src/geometry.rs

```diff
@@ -39,168 +39,235 @@ pub struct Geometry {
 }
 
 impl Geometry {
-    /// Create new triangle face geometry from provided faces and
-    /// vertices, and compute normals based on `normal_strategy`.
+    /// Creates new triangular geometry from provided triangle faces
+    /// and vertices, and computes normals based on `normal_strategy`.
     ///
     /// # Panics
     /// Panics if faces refer to out-of-bounds vertices.
-    pub fn from_triangle_faces_with_vertices_and_computed_normals(
-        faces: Vec<(u32, u32, u32)>,
-        vertices: Vertices,
+    pub fn from_triangle_faces_with_vertices_and_computed_normals<F, V>(
+        faces: F,
+        vertices: V,
         normal_strategy: NormalStrategy,
-    ) -> Self {
-        let mut normals = Vec::with_capacity(faces.len());
-        let vertices_range = 0..cast_u32(vertices.len());
-        for &(v1, v2, v3) in &faces {
-            assert!(
-                vertices_range.contains(&v1),
-                "Faces reference out of bounds position data"
-            );
-            assert!(
-                vertices_range.contains(&v2),
-                "Faces reference out of bounds position data"
-            );
-            assert!(
-                vertices_range.contains(&v3),
-                "Faces reference out of bounds position data"
-            );
+    ) -> Self
+    where
+        F: IntoIterator<Item = (u32, u32, u32)>,
+        V: IntoIterator<Item = Point3<f32>>,
+    {
+        // To avoid one additional cloning of this collection, we
+        // first materialize it into its final structure (Vec<Face>),
+        // and later assert the only kind of faces there are
+        // triangles.
+        let faces_collection: Vec<_> = faces
+            .into_iter()
+            .enumerate()
+            .map(|(i, (i1, i2, i3))| {
+                let normal_index = cast_u32(i);
+                TriangleFace::new_separate(i1, i2, i3, normal_index, normal_index, normal_index)
```

Comment:
This will definitely have to be fixed once we add `NormalStrategy::Smooth`, but I would prefer not to do it now, but rather when we implement the smooth strategy (so we can test the fixes are correct). Many other small implementation things will have to be revised once that happens.

For the record, this was already implemented for a while, the only reason it shows up on the diff is because I moved the function impls around.

Reply:
it seems inconsistent though - there is a separate logic for the normal strategy and not for normal creation. shouldn't all future features be treated the same way (either way)?

Reply:
Aaah, so that's what you mean by that. Ok, I can wrap it in the match statement if that makes it more consistent for you.

Do not though that probably the whole structure of this function will have to be reworked for smooth normals.

Reply:
exactly what i wanted to propose